File: AndroidCinema/app/src/main/java/com/crpr/androidcinema/data/api/factories/InterceptorsFactory.java
package com.crpr.androidcinema.data.api.factories;

import com.crpr.androidcinema.BuildConfig;
import com.crpr.androidcinema.data.api.interceptors.AuthInterceptor;
import com.crpr.androidcinema.data.api.interceptors.LogInterceptor;

import javax.inject.Provider;

import okhttp3.Interceptor;
import okhttp3.logging.HttpLoggingInterceptor;

/**
 * Created by cribeiro on 08/01/2016.
 */
public class InterceptorsFactory {

    public static final int LOG = 0;
    public static final int AUTH = 1;
    public static final int HTTP = 2;

    Provider<LogInterceptor> _logProvider;
    Provider<AuthInterceptor> _authProvider;
    Provider<HttpLoggingInterceptor> _httpInterceptor;

    public InterceptorsFactory(Provider<LogInterceptor> logProvider, Provider<AuthInterceptor> authProvider,
                               Provider<HttpLoggingInterceptor> httpProvider){
        this._logProvider = logProvider;
        this._authProvider = authProvider;
        this._httpInterceptor = httpProvider;
    }

    public Interceptor get(int type){
        switch (type){
            case LOG:
                return _logProvider.get();

            case AUTH:
                return _authProvider.get();

            case HTTP:
                HttpLoggingInterceptor interceptor = _httpInterceptor.get();
                interceptor.setLevel(BuildConfig.DEBUG ?
                        HttpLoggingInterceptor.Level.BODY :
                        HttpLoggingInterceptor.Level.BASIC);
                return interceptor;

            default:
                return null;
        }
    }
}

File: AndroidCinema/app/src/main/java/com/crpr/androidcinema/data/api/factories/InterceptorsFactory.kt
package com.crpr.androidcinema.data.api.factories

import com.crpr.androidcinema.BuildConfig
import com.crpr.androidcinema.data.api.interceptors.AuthInterceptor
import com.crpr.androidcinema.data.api.interceptors.LogInterceptor
import okhttp3.Interceptor
import okhttp3.logging.HttpLoggingInterceptor
import javax.inject.Provider

/**
 * Created by cribeiro on 08/01/2016.
 */
class InterceptorsFactory(internal var _logProvider: Provider<LogInterceptor>,
                          internal var _authProvider: Provider<AuthInterceptor>,
                          internal var _httpInterceptor: Provider<HttpLoggingInterceptor>) {

    operator fun get(type: Int): Interceptor? {
        when (type) {
            LOG -> return _logProvider.get()

            AUTH -> return _authProvider.get()

            HTTP -> {
                val interceptor = _httpInterceptor.get()
                interceptor.level = if (BuildConfig.DEBUG)
                    HttpLoggingInterceptor.Level.BODY
                else
                    HttpLoggingInterceptor.Level.BASIC
                return interceptor
            }

            else -> return null
        }
    }

    companion object {

        val LOG = 0
        val AUTH = 1
        val HTTP = 2
    }
}
